File: energycommunities/src/main/java/it/unical/demacs/asd/energycommunities/data/services/BatteryService.java
package it.unical.demacs.asd.energycommunities.data.services;

import java.util.List;

import it.unical.demacs.asd.energycommunities.dto.battery.BatteryDto;

public interface BatteryService {

    List<BatteryDto> get_batteries_by_plan(Long plan_id);

    BatteryDto get_battery(Long battery_id);

    void delete_battery(Long battery_id);

    BatteryDto add_battery(Long plan_id, BatteryDto battery);
}

File: energycommunities-ng/src/app/services/battery.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { BatteryDto } from "../model/battery/BatteryDto";
import { Observable } from "rxjs";

@Injectable({ providedIn: 'root' })
export class BatteryService {

    private readonly baseUrl = 'http://localhost:8080/battery'
    constructor(private readonly http: HttpClient) { }

    add_battery(plan_id: number, battery: BatteryDto): Observable<BatteryDto> {
        const battery_dto: BatteryDto = {
            id: null,
            plan: null,
            model: battery.model,
            capacity: battery.capacity,
            price: battery.price
        }

        return this.http.post<BatteryDto>(`${this.baseUrl}/plan/${plan_id}`, battery_dto);
    }

    get_battery(battery_id: number): Observable<BatteryDto> {
        return this.http.get<BatteryDto>(`${this.baseUrl}/${battery_id}`);
    }

    delete_battery(battery_id: number): Observable<void> {
        return this.http.delete<void>(`${this.baseUrl}/${battery_id}`);
    }

    get_batteries_by_plan(plan_id: number): Observable<BatteryDto[]> {
        return this.http.get<BatteryDto[]>(`${this.baseUrl}/all/${plan_id}`);
    }
}